CSharpMath/Structures/Dictionary.cs

```diff
@@ -0,0 +1,278 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+
+namespace CSharpMath.Structures {
+
+  [SuppressMessage("Naming", "CA1710:Identifiers should have correct suffix", Justification = NotACollection)]
+  [SuppressMessage("Design", "CA1010:Collections should implement generic interface", Justification = NotACollection)]
+  public class ProxyAdder<TKey, TValue> : IEnumerable {
```

Comment:
What is a ProxyAdder?  Needs documentation.

Reply:
Added

Reply:
This should add `// HACK`.
But then, why is a hack needed here?

Reply:
How is this a `HACK`?

Reply:
It is a hack because it involves implementing an interface for which there is no valid implementation! That is obvious.

I gather that the purpose (undocumented) of this is to initialize one of your special dictionaries.

The question is why is this hack needed. Can't you create an Array with the items you want and then pass it in to one of the special dictionaries you want programatically? Then you don't need all the special add methods and the ProxyAdder. This would result in a cleaner approach. Would the startup time involved be significant (e.g, more than 10ms)? I would guess it's sub-1ms since there are not millions of elements.

Reply:
>It is a hack because it involves implementing an interface for which there is no valid implementation! That is obvious.

This is to satisfy C#'s collection initializers' unreasonable requirement of `IEnumerable` despite not being used by the feature. Complain to C#.
>I gather that the purpose (undocumented) of this is to initialize one of your special dictionaries.

This is clearly documented at
```
const string NotACollection = "This is not a collection. It implements IEnumerable just to support collection initializers.";
```
which explains its purpose.

>The question is why is this hack needed. Can't you create an Array with the items you want and then pass it in to one of the special dictionaries you want programatically? Then you don't need all the special add methods and the ProxyAdder. This would result in a cleaner approach.

What type would such an array be? Each Add is variable-length. Including a special-purpose structure conflicts with your "lines of code added" measure and is non-idiomatic by not using collection initializers. Seems like you are suggesting a larger hack to cover this "hack".

>Would the startup time involved be significant (e.g, more than 10ms)? I would guess it's sub-1ms since there are not millions of elements.

And on top of that we also need to measure the time. A lot of additional work for more non-idiomatic code.